### PR TITLE
Implement Fuzzy Search

### PR DESCRIPTION
**Pull request by TheRealRavager**
Implement Fuzzy Search. ~~Only merge after PR #181 is merged~~ PR #181 is merged.

### CODE REVIEW

**Comment by TheRealRavager** [comment]
Please review my code first. Right now fuzzy find is its own command called using `fuzfind` but ill change that soon

**Comment by WilliamRyank** [comment]
Do we need 2 Find Command Class? or make normal find fuzzy? whats the idea of separating normal find and fuzzy

**Comment by TheRealRavager** [comment]
> Do we need 2 Find Command Class? or make normal find fuzzy? whats the idea of separating normal find and fuzzy

The idea is that `find` is fuzzy while `exactFind` reduces clutter by only displaying exact matches. I'll change the class names to reflect this

File: src/main/java/organice/logic/commands/ExactFindCommand.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    public ExactFindCommand(PersonContainsPrefixesPredicate predicate) {
+        requireNonNull(predicate);
+        this.predicate = predicate;
+    }
+
+    @Override
+    public CommandResult execute(Model model) {
+        requireNonNull(model);
+        model.updateFilteredPersonList(predicate);
```
Comment (C-likethis123) [inline src/main/java/organice/logic/commands/ExactFindCommand.java]: You may need to add `model.updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS)` to reset the FilteredList before applying your Predicate.

Comment (TheRealRavager) [reply src/main/java/organice/logic/commands/ExactFindCommand.java]: updating the predicate does not change the total number of items inside the filtered persons. FilteredList already takes care of the situation you are thinking of

File: src/main/java/organice/logic/commands/FindCommand.java [hunk 2]
```diff
@@ -1,44 +1,154 @@
...
+     */
+    private List<Person> fuzzyMatch(ArgumentMultimap argMultimap, List<Person> inputList) {
+        // Fuzzy Match by Levenshtein Distance is not implemented for following prefixes:
+        // Age, Priority, BloodType, TissueType
+        // Typos in these fields have a very small Levenshtein Distance (LD) as typical field length is very small
+
+        List<String> nameKeywords = argMultimap.getAllValues(PREFIX_NAME);
+        List<String> nricKeywords = argMultimap.getAllValues(PREFIX_NRIC);
+        List<String> phoneKeywords = argMultimap.getAllValues(PREFIX_PHONE);
+        List<String> typeKeywords = argMultimap.getAllValues(PREFIX_TYPE);
```
Comment (C-likethis123) [inline src/main/java/organice/logic/commands/FindCommand.java]: As discussed earlier, you should type validate.

File: src/main/java/organice/logic/commands/FindCommand.java [hunk 3]
```diff
@@ -1,44 +1,154 @@
...
     }
 
     @Override
     public CommandResult execute(Model model) {
         requireNonNull(model);
-        model.updateFilteredPersonList(predicate);
-        return new CommandResult(
-                String.format(Messages.MESSAGE_PERSONS_LISTED_OVERVIEW, model.getFilteredPersonList().size()));
+
+        List<Person> allPersons = Arrays.asList(model.getFilteredPersonList().toArray(Person[]::new));
```
Comment (C-likethis123) [inline src/main/java/organice/logic/commands/FindCommand.java]: To get all persons in ORGANice, the initial predicate should be `PREDICATE_SHOW_ALL_PERSONS`.

File: src/main/java/organice/logic/parser/ExactFindCommandParser.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ */
+public class ExactFindCommandParser implements Parser<ExactFindCommand> {
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the ExactFindCommand
+     * and returns a ExactFindCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
+    public ExactFindCommand parse(String args) throws ParseException {
+        args = args.replaceAll("\n", " ").replaceAll("\\s+", " ");
```
Comment (WilliamRyank) [inline src/main/java/organice/logic/parser/ExactFindCommandParser.java]: What is the purpose of `replaceAll("\n")`? i think trimmed has deal with multi spaces in between right? no need to replace all `\\s+`. Even i think how the tokenizer work also they remove extra spaces. Need to check

File: src/main/java/organice/logic/commands/FindCommand.java (hunk 2 above)
Comment (TheRealRavager) [reply src/main/java/organice/logic/commands/FindCommand.java]: Another PR

File: src/main/java/organice/logic/commands/FindCommand.java (hunk 3 above)
Comment (TheRealRavager) [reply src/main/java/organice/logic/commands/FindCommand.java]: Created new method in `ModelManager` to handle this explicit behaviour.

File: src/main/java/organice/logic/parser/ExactFindCommandParser.java (hunk 4 above)
Comment (TheRealRavager) [reply src/main/java/organice/logic/parser/ExactFindCommandParser.java]: Replacing newlines is a defensive action on my side of the code. Replacing `\\s+` with `' '` takes care of whitespaces in between individual words (trim only removes preceding and trailing whitespace). Have not tested tokenize, but a glace at the code suggests it doesnt protect against whitespace in between words (I may be wrong). Either way i think its better to be defensive

**Pull request merged**